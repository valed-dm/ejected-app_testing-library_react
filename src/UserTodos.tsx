import React, { FC, useState, useEffect } from "react";
import { getUser } from "./App"; 

interface UserTodosProps {
    userName: string;
}

const UserTodos: FC<UserTodosProps> = ({ userName }) => {
  const [msg, setMsg] = useState("");
  const [todos, setTodos] = useState<Array<JSX.Element>>();

  const setUserTodos = async () => {
    const user = await getUser(userName);
    setMsg(`Welcome to React testing, ${user.name}!`);
    console.log("user by username", user);

    const todosResponse = await fetch(
      "https://jsonplaceholder.typicode.com/todos"
    );
    if (todosResponse.ok) {
      const todos = await todosResponse.json();
      const userTodos = todos.filter((todo: any) => {
        return todo.userId === user.id;
      });
      const todoList = userTodos.map((todo: any) => {
        return <li key={todo.id}>{todo.title}</li>;
      });

      setTodos(todoList);
      console.log("user todos", userTodos);
    }
  };

  useEffect(() => {
    if (userName) {
      setUserTodos();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userName]);

  return (
    <div>
      <div>
        <label data-testid="final-msg">{msg}</label>
      </div>
      <ul style={{ marginTop: "1rem", listStyleType: "none" }}>{todos}</ul>
    </div>
  );
};

export default React.memo(UserTodos);
