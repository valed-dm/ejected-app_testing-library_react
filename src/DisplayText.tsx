import React, { useState, FC } from "react";
import UserTodos from "./UserTodos";

const DisplayText: FC = () => {
  const [txt, setTxt] = useState("");
  const [todoControl, setTodoControl] =
    useState<ReturnType<typeof UserTodos>>();

  const onChangeTxt = (e: React.ChangeEvent<HTMLInputElement>) => {
    setTxt(e.target.value);
  };
  const onClickShowMsg = async (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();

    setTodoControl(null);
    setTodoControl(<UserTodos userName={txt} />);
  };

  return (
    <form>
      <div>
        <label>Enter your name</label>
      </div>
      <div>
        <input data-testid="user-input" value={txt} onChange={onChangeTxt} />
      </div>
      <div>
        <button data-testid="input-submit" onClick={onClickShowMsg}>
          Show Message
        </button>
      </div>
      {todoControl}
    </form>
  );
};

export default DisplayText;
