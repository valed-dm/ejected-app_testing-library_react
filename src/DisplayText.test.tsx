import {
  render,
  screen,
  fireEvent,
  cleanup,
  waitFor,
} from "@testing-library/react";
import DisplayText from "./DisplayText";
import UserTodos from "./UserTodos";
import "@testing-library/jest-dom/extend-expect";
afterEach(cleanup);

describe("Test DisplayText", () => {
  const userName = "testuser";
  const userFullName = "Dmitrii Tester";
  const getUserMock = (userName: string): jest.Mock<Promise<any>> => {
    const promise = new Promise<any>((res, rej) => {
      res({ name: userFullName });
    });
    const getUserFullName = jest.fn(async (userName: string): Promise<any> => {
      return promise;
    });
    return getUserFullName;
  };

  it("renders without crashing", () => {
    const { baseElement } = render(<DisplayText />);
    console.log(baseElement.innerHTML);
    expect(baseElement).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { baseElement } = render(<DisplayText />);
    expect(baseElement).toMatchSnapshot();
  });

  it("receives input text", () => {
    render(<DisplayText />);
    const input = screen.getByTestId("user-input");
    fireEvent.change(input, { target: { value: userName } });
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(userName);
  });

  it("shows welcome message", async () => {
    //if getUserFullName is removed from here this test fails
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const getUserFullName = getUserMock(userName);

    const msg = `Welcome to React testing, ${userFullName}!`;
    render(
      <>
        <DisplayText />
        <UserTodos userName={userName} />
      </>
    );
    const input = screen.getByTestId("user-input");
    const label = screen.getByTestId("final-msg");
    fireEvent.change(input, { target: { value: userName } });
    const btn = screen.getByTestId("input-submit");
    fireEvent.click(btn);

    expect(label).toBeInTheDocument();
    await waitFor(() => {
      expect(label.innerHTML).toBe(msg);
    });
  });
});
