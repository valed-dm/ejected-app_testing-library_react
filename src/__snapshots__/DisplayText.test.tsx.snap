// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DisplayText matches snapshot 1`] = `
<body>
  <div>
    <form>
      <div>
        <label>
          Enter your name
        </label>
      </div>
      <div>
        <input
          data-testid="user-input"
          value=""
        />
      </div>
      <div>
        <button
          data-testid="input-submit"
        >
          Show Message
        </button>
      </div>
    </form>
  </div>
</body>
`;
